### DISCUSSION TITLE
Support non-JSON cookies in Remix cookie APIs

### DISCUSSION THREAD

**Discussion by nobleach** [remix-run/remix#7922] [Proposals]
### What version of Remix are you using?

2.2.0

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

I've created the following cookie:
```typescript
export const pgAdminCookie = createCookie('PGADMIN_LANGUAGE', {
  maxAge: 604_800, // one week
});
```
When attempting to parse it and pull a value from it, I get `null`. 

```typescript
  const cookieHeader = request.headers.get('Cookie');
  console.debug('DEBUGPRINT[2]: upsell.tsx:17: cookieHeader=', cookieHeader);
  const cookie = await pgAdminCookie.parse(cookieHeader);
  console.debug('DEBUGPRINT[1]: upsell.tsx:18: cookie=', cookie);
```

I can see via that first console log (of all Cookie headers) that the cookie is indeed in the header:

```sh
DEBUGPRINT[2]: upsell.tsx:17: cookieHeader= PGADMIN_LANGUAGE=en;
```

### Expected Behavior

I expect that second console.log (of the parsed cookie) to log `en`.

### Actual Behavior

It logs `null`.

The same is true for any cookie I try to log. They all show up in `request.headers.get('Cookie');`, but all parse to `null`;

Perhaps some weird async thing? My loader is async, I am awaiting the parse.

**Comment by nobleach** [comment]
Looks like all cookies are expected to be `JSON.parse()`-able. Otherwise `decodeCookieValue()` ([here](https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/cookies.ts#L188)):

```typescript
    return JSON.parse(decodeURIComponent(myEscape(atob(value))));
```

will throw. Perhaps we should check if the cookie contains something that resembles JSON? Cookies often contain strings, or encoded tokens/etc.

**Comment by brophdawg11** [comment]
I'm going to convert this to a Proposal since it's a feature request - Remix cookies today are JSON objects so they're a bit easier to store multiple pieces of information in.  I think you have 2 options at the moment:

* Use something like https://www.npmjs.com/package/cookie to parse non-Remix cookies
* Migrate non-Remix cookies to Remix cookies so the cookie becomes something like `{lang: 'en'}`
  * You could do this in conjunction with option 1 as well and your loader could "move" the cookie to a Remix cookie if it detected a cookie of the old format

**Reply by sergiodxa** [reply]
You don't need to store an object inside a Remix cookie, you can store plain values, the issue is the base64 encoding, Remix expects the value to be encoded so it can decode it, and if that fails it will return `null`.

Where @nobleach cookie seems to be the plain value directly `PGADMIN_LANGUAGE=en`.

**Reply by brophdawg11** [reply]
I think it needs to be `JSON.parse`-able, no?  So you could technically store a string _with the quotes_ (`"hi"`) but storing just a string (like `en`, even properly encoded) will throw on `JSON.parse` and return an empty object? 

https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/cookies.ts#L188

**Comment by nobleach** [comment]
@brophdawg11 I'm sure I missed the part where Remix cookies differed from just regular cookies. Since I don't have control over all the various cookies set/read by my org, I went with the first option. I probably should have understood that `createCookie` wasn't just creating a wrapper object, but expecting to actually be in charge of the cookie lifecycle.

If this is the plan going forward, I'm sure some docs about working with existing cookies is sufficient.

**Reply by itoro-sh** [reply]
Thanks for bringing this one up. I'm having the same problem right now, when trying to migrate an old project with old cookies created by services we need to reuse. I also thought that createCookie was a wrapper to expose get/set functionalities and whatnot and that the value you see on the cookie headers object was the one you get on the parse. Quite confusing 🤔

**Comment by ashitikov** [comment]
Hello! Faced the same problem while implementing i18next support with detection of language from cookies, and the current workaround is:
```
export const lngCookie = createCookie("i18next", {
  decode: (value) => {
    return btoa(JSON.stringify(value));
  },
});
```

It's worth to mention, that decode function being executed for every cookie value. If you have multiple values, you need to check type of value, to make sure you are using correct cookie.
